**MAESTRÍA EN INTELIGENCIA ARTIFICIAL APLICADA**

**Curso: TC5053 - Ciencia y analítica de datos**

Tecnológico de Monterrey

Prof Grettel Barceló Alonso

**Semana 3**
Bases, almacenes y manipulación de datos

---

*   NOMBRE: IARA PÉREZ PINEDA
*   MATRÍCULA: A01705855


---

En esta actividad usarás una base de datos relacional basada en el informe de participación y la lista del top 10 de Netflix. Incluye películas y programas de televisión, así como información sobre temporadas, métricas de visualización, fechas de estreno, duración y más, organizada en las siguientes tablas:

* `movie`: Información general de las películas.

* `tv_show`: Información general de los programas de televisión.

* `season`: Datos de las temporadas asociadas a cada programa de TV.

* `view_summary`: Métricas de visualización y rendimiento de películas o temporadas.

Revisa con detalle su esquema para que comprendas cómo se relacionan las tablas anteriores.

**NOTA IMPORTANTE:** Asegúrate de responder *explícitamente* todos los cuestionamientos.


`PyMySQL` es una librería escrita en Python puro que funciona como conector (*driver*) para motores de bases de datos MySQL, permitiendo abrir conexiones, ejecutar consultas SQL y recuperar resultados directamente desde programas en Python.

In [83]:
pip install pymysql

`SQLAlchemy` es una librería de Python que facilita la interacción con bases de datos y permite mantener un pool de conexiones eficiente, gestionar *commits* y *rollbacks* de forma automática y asegurar que múltiples conexiones simultáneas se manejen de manera segura, incluso cuando se ejecutan consultas SQL “puras”

In [84]:
# Importa las librerías necesarias
import pymysql
import sqlalchemy as sqla
import pandas as pd

Se crea una conexión (`conn`) para luego invocar declaraciones SQL.

In [85]:
# motor+driver://usuarioBD:clave@ipHostDBMS:puerto/esquema
# pool_recycle controla el tiempo máximo de vida de una conexión en el pool (3600 segundos = 1 hora)
db = sqla.create_engine('mysql+pymysql://mnaTC5053User:mnaTC5053Pass!@20.127.169.219:3306/Netflix', pool_recycle=3600)
conn = db.connect()

Para que tus consultas sean más legibles y fáciles de mantener, puedes usar este formato multilínea con comillas triples y `sqla.text()`. Por ejemplo:

```
query = sqla.text("""
  SELECT ---
  FROM ---
  WHERE ---
""")
pd.read_sql(query, conn)
```

1.	Extrae toda la información de las películas que duran más de 5 horas.

In [86]:
query = sqla.text("""
SELECT *
FROM movie
WHERE runtime > 300
""")

pd.read_sql(query, conn)

,id,created_date,modified_date,available_globally,locale,original_title,release_date,runtime,title
0,5793,2024-01-01,2024-01-01,b'\x00',None,日本統一シリーズ: 映画シリーズ,None,3892,Nihontouitsu Series: Film Series
1,5794,2024-01-01,2024-01-01,b'\x00',None,釣りバカ日誌: 映画シリーズ,None,2120,Free and Easy Series: Film Series
2,5874,2024-01-01,2024-01-01,b'\x01',None,None,2021-08-06,312,Navarasa: Limited Series
3,9729,2024-01-01,2024-01-01,b'\x00',None,織田同志会 織田征仁: 映画シリーズ,None,710,Seiji Oda: Film Series
4,9730,2024-01-01,2024-01-01,b'\x00',None,キングダム～首領になった男～: 映画シリーズ,None,427,Kingdom ~ The Man Who Became the Top ~: Film S...


2.	¿Cuál es el porcentaje de películas disponibles únicamente en EU en relación con el total, excluyendo los valores `NULL`? La consulta SQL debe entregar directamente el porcentaje final. No se deben devolver resultados parciales para realizar el cálculo en Python.

In [87]:
query = sqla.text("""
SELECT
    100.0 * SUM(CASE WHEN available_globally = 0 THEN 1 ELSE 0 END)
    / COUNT(available_globally) AS porcentaje_solo_EU
FROM movie
""")

pd.read_sql(query, conn)

,porcentaje_solo_EU
0,83.57175


3.	¿Cuáles son los idiomas o regiones originales en la tabla de películas?
* ¿Cuántos registros tienen el campo `locale` con valor `NULL`? (NULL en SQL ⇔ None en Python)

In [88]:
query = sqla.text("""
SELECT DISTINCT locale
FROM movie
ORDER BY locale
""")

pd.read_sql(query, conn)

,locale
0,None
1,en


In [89]:
query = sqla.text("""
SELECT COUNT(*) AS registros_locale_null
FROM movie
WHERE locale IS NULL
""")

pd.read_sql(query, conn)

,registros_locale_null
0,11255


4.	Asumiendo que los valores `NULL` en `locale` corresponden a otro idioma (diferente del inglés), el título original de la película NO debería coincidir con el título principal en dichos registros.
* Determina cuántas películas tienen títulos diferentes en estos dos campos (`title` y `original_title`).
*  ¿Coinciden los resultados (cantidad de `NULL` y títulos diferentes)? Si no es así, identifica qué características tienen los registros restantes.
* Finalmente, concluye si la suposición de que los valores `NULL` en `locale` indican que la película está en otro idioma es válida.

In [90]:
query = sqla.text("""
SELECT COUNT(*) AS titulos_diferentes
FROM movie
WHERE title <> original_title
""")

pd.read_sql(query, conn)

,titulos_diferentes
0,3947


In [91]:
query = sqla.text("""
SELECT
    COUNT(*) AS cantidad,
    CASE
        WHEN original_title IS NULL THEN 'original_title NULL'
        WHEN title = original_title THEN 'titulos iguales'
        ELSE 'titulos diferentes'
    END AS caracteristica
FROM movie
WHERE locale IS NULL
GROUP BY caracteristica
""")

pd.read_sql(query, conn)

,cantidad,caracteristica
0,7291,original_title NULL
1,3947,titulos diferentes
2,17,titulos iguales


Los resultados no coinciden directamente porque existen 11,255 registros con locale = NULL, pero solo 3,947 de esos presentan títulos diferentes. De los registros que restan, 7,291 también tienen original_title = NULL y 17 presentan el mismo título en title y original_title. Por eso, no es válido asumir que un valor NULL en locale necesariamente indica que la película está en un idioma distinto del inglés, pues en la mayoría de estos casos no existe un título original con el cual comprobarlo.

5.	Determina el título de la película que ha permanecido más tiempo en el top 10.

In [92]:
query = sqla.text("""
SELECT
    movie.title,
    view_summary.cumulative_weeks_in_top10
FROM movie
JOIN view_summary
    ON movie.id = view_summary.movie_id
ORDER BY view_summary.cumulative_weeks_in_top10 DESC
LIMIT 1
""")

pd.read_sql(query, conn)

,title,cumulative_weeks_in_top10
0,The Boss Baby,22


6.	Identifica los 5 programas de TV con mayor cantidad de temporadas.

In [93]:
query = sqla.text("""
SELECT
    tv_show.title,
    COUNT(season.id) AS cantidad_temporadas
FROM tv_show
JOIN season
    ON tv_show.id = season.tv_show_id
GROUP BY tv_show.id, tv_show.title
ORDER BY cantidad_temporadas DESC
LIMIT 5
""")

pd.read_sql(query, conn)

,title,cantidad_temporadas
0,Grey's Anatomy,20
1,Naruto Shippuden,20
2,Heartland (2007),17
3,It's Always Sunny in Philadelphia,16
4,NCIS,15


7.	¿Cuáles son los intervalos de fechas de los resúmenes en la tabla `view_summary` de los períodos (`duration`) semestrales?

In [94]:
query = sqla.text("""
SELECT DISTINCT
    start_date,
    end_date
FROM view_summary
WHERE duration = 'SEMI_ANNUALLY'
ORDER BY start_date
""")

pd.read_sql(query, conn)

,start_date,end_date
0,2023-07-01,2023-12-31
1,2024-01-01,2024-06-30


In [95]:
conn.close()
conn = db.connect()

8.	Ordena las temporadas de *Grey's Anatomy* según la cantidad de vistas registradas en el primer período semestral de 2024.
* ¿Cómo interpretarías los resultados obtenidos?

In [96]:
query = sqla.text("""
SELECT
    season.season_number,
    season.title,
    view_summary.views
FROM tv_show
JOIN season
    ON tv_show.id = season.tv_show_id
JOIN view_summary
    ON season.id = view_summary.season_id
WHERE tv_show.title = 'Grey''s Anatomy'
    AND view_summary.duration = 'SEMI_ANNUALLY'
    AND view_summary.start_date = '2024-01-01'
    AND view_summary.end_date = '2024-06-30'
ORDER BY view_summary.views DESC
""")

pd.read_sql(query, conn)

,season_number,title,views
0,1,Grey's Anatomy: Season 1,3600000
1,2,Grey's Anatomy: Season 2,3100000
2,3,Grey's Anatomy: Season 3,2900000
3,5,Grey's Anatomy: Season 5,2900000
4,4,Grey's Anatomy: Season 4,2900000
5,6,Grey's Anatomy: Season 6,2800000
6,7,Grey's Anatomy: Season 7,2700000
7,8,Grey's Anatomy: Season 8,2600000
8,9,Grey's Anatomy: Season 9,2500000
9,10,Grey's Anatomy: Season 10,2400000


Las temporadas iniciales de Grey's Anatomy presentan la mayor cantidad de vistas durante el primer semestre del 2024. La temporada 1 encabeza los resultados con 3,600,000 vistas, mientras que la temporada 20 registra la menor cantidad con 100,000. En general se observa una disminución de las vistas conforme aumenta el número de temporada, aunque existen algunos empates y pequeñas variaciones entre temporadas intermedias.

9.	Todas las consultas anteriores podrían hacerse también con la lógica relacional implementada en Pandas, que permite replicar la mayoría de las operaciones de SQL. Si los dataframes se han llenado como sigue, resuelve la consulta 8 utilizando las funciones de Pandas.

In [97]:
tv_show = pd.read_sql(sqla.text('SELECT * FROM tv_show'), conn)
season = pd.read_sql(sqla.text('SELECT * FROM season'), conn)
view_summary = pd.read_sql(sqla.text('SELECT * FROM view_summary'), conn)

In [98]:
resultado = tv_show.merge(
    season,
    left_on='id',
    right_on='tv_show_id',
    suffixes=('_show', '_season')
)

resultado = resultado.merge(
    view_summary,
    left_on='id_season',
    right_on='season_id'
)

resultado = resultado[
    (resultado['title_show'] == "Grey's Anatomy") &
    (resultado['duration'] == 'SEMI_ANNUALLY') &
    (resultado['start_date'].astype(str) == '2024-01-01') &
    (resultado['end_date'].astype(str) == '2024-06-30')
]

resultado[
    ['season_number', 'title_season', 'views']
].sort_values('views', ascending=False)

,season_number,title_season,views
2309,1.0,Grey's Anatomy: Season 1,3600000
2311,2.0,Grey's Anatomy: Season 2,3100000
2313,3.0,Grey's Anatomy: Season 3,2900000
2315,5.0,Grey's Anatomy: Season 5,2900000
2317,4.0,Grey's Anatomy: Season 4,2900000
2319,6.0,Grey's Anatomy: Season 6,2800000
2321,7.0,Grey's Anatomy: Season 7,2700000
2323,8.0,Grey's Anatomy: Season 8,2600000
2325,9.0,Grey's Anatomy: Season 9,2500000
2327,10.0,Grey's Anatomy: Season 10,2400000


`MySQL` es un sistema de gestión de bases de datos relacional, pero desde Python también es posible extraer información de bases de datos no relacionales, como `Firestore`, `MongoDB` o `Cassandra`, utilizando conectores o integraciones específicas. Esto permite combinar datos de diferentes fuentes según las necesidades del análisis o la aplicación.

En el siguiente ejercicio usarás un ejemplo con `Firestore` desde Python. Para ello utilizarás los módulos `credentials` y `firestore` de la biblioteca `firebase_admin`.

In [99]:
import firebase_admin
from firebase_admin import credentials
from firebase_admin import firestore

En `Firestore`, a diferencia de `MySQL` donde se utiliza un usuario y contraseña para conectarse, la autenticación se realiza mediante un archivo JSON que almacena las credenciales necesarias para acceder a la base de datos. Este archivo contiene las claves y la información de configuración que permiten a Python establecer la conexión de manera segura.

In [100]:
cred = credentials.Certificate('consultancy.json')
firebase_admin.initialize_app(cred)
db = firestore.client()

In [101]:
# consultancy.json almacena la clave privada para autenticar una cuenta y autorizar el acceso a los servicios
# A través de la función Certificate(), se regresa una credencial inicializada, que puedes utilizar para crear una nueva instancia de la aplicación


10.	Investiga cómo leer la colección `EMPLOYEE` y mostrar su contenido en un dataframe. Asegúrate de incluir el `id` en el resultado

In [102]:
empleados = []

for doc in db.collection('EMPLOYEE').stream():
    datos = doc.to_dict()
    datos['id'] = doc.id
    empleados.append(datos)

df_empleados = pd.DataFrame(empleados).set_index('id')
df_empleados

,emp_fname,emp_lname,emp_hiredate
id,,,
8LcLuxVHGAd2d9IQc5jF,David,Senior,1989-07-12 06:00:00+00:00
Fzd60D6Z2CU4n0wVV8YN,William,Smithfield,2004-06-04 05:00:00+00:00
lX5xuQ5w3i6ib2ExccWY,John,News,2000-11-08 06:00:00+00:00
yocFj2lichOkbAj9NBfp,June,Arbough,1996-12-01 06:00:00+00:00


In [103]:
firebase_admin.delete_app(firebase_admin.get_app())

---

**Declaración de uso de IA**

OpenAI. (2026). ChatGPT (GPT-5.6 Sol) [Modelo de lenguaje grande], utilizado como apoyo para la generación, revisión y depuración de código SQL y Python/Pandas durante el desarrollo de la actividad.